---
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: show
    highlight: zenburn
    theme: flatly
---

```{r setup-knitr, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
                      cache=TRUE, cache.lazy=FALSE, highlight=TRUE, autodep=TRUE,
                      warning=FALSE, error=FALSE,
                      message=FALSE, prompt=TRUE, comment='', fig.cap='')
```

We created a tidy summary of MISO results for retained intron events
generated from the Ensemble gene annotation for GRCh37 version 68.

```{r load-misori}
library(readr)
library(dplyr)
library(tidyr)
library(ggplot2)
library(cowplot)
library(viridis)
misofn = "miso/combined-miso.tsv"
miso = read_tsv(misofn) %>%
  tidyr::separate(samplename, c("genotype", "replicate"), sep="_")
events2gene = read_tsv("miso/event-to-gene.tsv", col_names=FALSE)
colnames(events2gene) = c("event_name", "symbol")
```

I grouped the samples by genotype and calculated the mean PSI value for
each event. I kept only events where a PSI value was calculated for an
event in all genotypes. Then I plotted the mean PSI value for the various
genotypes against each other, similar to figure 3C in the paper that was
linked.

# PSI plots

```{r summarise}
misosum = miso %>%
  dplyr::group_by(eventtype, event_name, genotype) %>%
  dplyr::summarise(mpsi=mean(miso_posterior_mean)) %>%
  tidyr::spread(genotype, mpsi)
misosum = misosum[complete.cases(misosum),]
misosum = misosum %>%
  mutate(prpf8_diff = ifelse(control - prpf8 > 0.10, "lower",
                      ifelse(control - prpf8 < -0.10, "higher", "same"))) %>%
  mutate(prpf38_diff = ifelse(control - prpf38 > 0.10, "lower",
                       ifelse(control - prpf38 < -0.10, "higher", "same")))
```

## PRPF8
```{r prpf8-ri}
ggplot(misosum, aes(control, prpf8, color=prpf8_diff)) +
  facet_wrap(~eventtype) +
  geom_point(alpha=0.3, size=1) +
  geom_abline() +
  xlab("\u03A8, control") +
  ylab("\u03A8, PRPF8") +
  scale_color_viridis(discrete=TRUE) +
  panel_border() + labs(color="") +
  guides(colour = guide_legend(override.aes = list(alpha = 1)))
```

## PRPF38
```{r prpf38-ri}
ggplot(misosum, aes(control, prpf38, color=prpf38_diff)) +
  facet_wrap(~eventtype) +
  geom_point(alpha=0.3, size=1) +
  geom_abline() +
  xlab("\u03A8, control") +
  ylab("\u03A8, PRPF38") +
  scale_color_viridis(discrete=TRUE) +
  panel_border() + labs(color="") +
  guides(colour = guide_legend(override.aes = list(alpha = 1)))
```

# Wilcoxon tests

Here is a summary of the counts of events for being higher, lower or the same
in PRPF8 compared to control, where higher and lower are different by 10% PSI.

## Retained intron (RI)

### PRPF8

```{r prf8-table}
risum = subset(misosum, eventtype == "RI")
table(risum$prpf8_diff)
wilcox.test(risum$control, risum$prpf8, alternative="less")
```

### PRPF38

```{r prpf38-table}
table(risum$prpf38_diff)
wilcox.test(risum$control, risum$prpf38, alternative="less")
```

### Table download
```{r write-tables}
risum = risum %>% left_join(events2gene, by="event_name")
write.table(risum, file="RI_summary.csv", row.names=FALSE, col.names=TRUE, quote=FALSE, sep=",")
```

[summary-table](RI_summary.csv)

## Skipped exon (SE)

### PRPF8
```{r prpf8-se-table}
sesum = subset(misosum, eventtype == "SE")
table(sesum$prpf8_diff)
wilcox.test(sesum$control, sesum$prpf8, alternative="less")
```

### PRPF38

```{r prpf38-se-table}
table(sesum$prpf38_diff)
wilcox.test(sesum$control, sesum$prpf38, alternative="less")
```

### Table download
```{r write-se-table}
sesum = sesum %>% left_join(events2gene, by="event_name")
write.table(sesum, file="SE_summary.csv", row.names=FALSE, col.names=TRUE, quote=FALSE, sep=",")
```

[summary-table](SE_summary.csv)

## Alternative 3' start site (A3SS)

```{r prpf8-a3ss-table}
a3sssum = subset(misosum, eventtype == "A3SS")
table(a3sssum$prpf8_diff)
wilcox.test(a3sssum$control, a3sssum$prpf8, alternative="less")
```

```{r prpf38-a3ss-table}
table(a3sssum$prpf38_diff)
wilcox.test(a3sssum$control, a3sssum$prpf38, alternative="less")
```

```{r write-a3ss-table}
a3sssum = a3sssum %>% left_join(events2gene, by="event_name")
write.table(a3sssum, file="A3SS_summary.csv", row.names=FALSE, col.names=TRUE, quote=FALSE, sep=",")
```

[summary-table](A3SS_summary.csv)

## Alternative 5' start site (A5SS)

```{r prpf8-a5ss-table}
a5sssum = subset(misosum, eventtype == "A5SS")
table(a5sssum$prpf8_diff)
wilcox.test(a5sssum$control, a5sssum$prpf8, alternative="less")
```

```{r prpf38-a5ss-table}
table(a5sssum$prpf38_diff)
wilcox.test(a5sssum$control, a5sssum$prpf38, alternative="less")
```

```{r write-a5ss-table}
a5sssum = a5sssum %>% left_join(events2gene, by="event_name")
write.table(a5sssum, file="A5SS_summary.csv", row.names=FALSE, col.names=TRUE, quote=FALSE, sep=",")
```

[summary-table](A5SS_summary.csv)

## Mutually exclusive exons (MXE)

```{r prpf8-mxe-table}
mxesum = subset(misosum, eventtype == "MXE")
table(mxesum$prpf8_diff)
wilcox.test(mxesum$control, mxesum$prpf8, alternative="less")
```

```{r prpf38-mxe-table}
table(mxesum$prpf38_diff)
wilcox.test(mxesum$control, mxesum$prpf38, alternative="less")
```

```{r write-mxe-table}
mxesum = mxesum %>% left_join(events2gene, by="event_name")
write.table(mxesum, file="MXE_summary.csv", row.names=FALSE, col.names=TRUE, quote=FALSE, sep=",")
```

[summary-table](MXE_summary.csv)

# Raw data

[raw-table](miso/combined-miso.tsv.gz)
